sources/advanced/Cli.ts

```diff
@@ -172,6 +177,25 @@ export type MiniCli<Context extends BaseContext> = CliOptions & {
   usage(command?: CommandClass<Context> | Command<Context> | null, opts?: {detailed?: boolean, prefix?: string}): string;
 };
 
+export type MandatoryContextKeys<Context extends BaseContext> = keyof MandatoryContext<Context>;
+
+export type MandatoryContext<Context extends BaseContext> = {
+  [K in Exclude<keyof Context, keyof BaseContext> as undefined extends Context[K] ? never : K]: Context[K];
+};
+
+export type UserProvidedContext<Context extends BaseContext> =
+  & MandatoryContext<Context>
+  & Partial<Omit<Context, MandatoryContextKeys<Context>>>;
+
+export type MaybeProvidedContext<Context extends BaseContext> =
+  MandatoryContextKeys<Context> extends never
+    ? {context?: UserProvidedContext<Context>}
+    : {context: UserProvidedContext<Context>};
+
+export type ProcessOptions<Context extends BaseContext> =
+    & MaybeProvidedContext<Context>
+    & {input: Array<string>, partial?: boolean};
```

Comment:
I'd mark `partial` as experimental since there's a great chance some things might need to be changed around it for completions.